internal/server/singleprocess/state/snapshot.go

```diff
@@ -0,0 +1,462 @@
+		return err
+	}
+
+	// We do want to check that this appears valid because if we can't open
+	// the file then the server will never start again until it is cleaned up.
+	log.Info("validating restore file")
+	sr, closer, err := snapshotReader(ri.StageTempPath, nil)
+	if err != nil {
+		log.Error("error while validating restore file", "err", err)
+		return fmt.Errorf("error validating restore data: %s", err)
+	}
+
+	var header pb.Snapshot_Header
+	err = sr.ReadMsg(&header)
+	if cerr := closer(); cerr != nil {
+		err = cerr
+	}
+	if err != nil {
+		log.Error("error while validating restore file", "err", err)
+		return fmt.Errorf("error validating restore data: %s", err)
+	}
+
+	// Replace our file
+	log.Info("atomically replacing file", "src", ri.StageTempPath, "dest", ri.StagePath)
+	if err := atomic.ReplaceFile(ri.StageTempPath, ri.StagePath); err != nil {
+		log.Error("error replacing file", "err", err)
+		return err
+	}
+
+	// Open the new database
+	log.Warn("snapshot staged for restore on next startup")
+	return nil
+}
+
+// finalizeRestore checks for any staged restore and moves it into place.
+// This will error if it fails for any reason which may prevent startup
+// but we have to prevent startup because the user wanted a restore.
+func finalizeRestore(log hclog.Logger, db *bolt.DB) (*bolt.DB, error) {
+	log.Debug("checking if DB restore is requested")
+	ri := newRestoreInfo(log, db)
+	if err := ri.Lock(); err != nil {
+		return db, err
+	}
+	defer ri.Unlock()
+
+	_, err := os.Stat(ri.StagePath)
+	if os.IsNotExist(err) {
+		log.Debug("no restore file found, no DB restore requested")
+		return db, nil
+	}
+	if err != nil {
+		log.Error("error checking for restore file", "err", err)
+		return db, err
+	}
+
+	log.Warn("restore file found, will initiate database restore")
+
+	// The file on disk should be in the form of pb.Snapshot. So what we need
+	// to do next is parse out the header to determine the format, and then
+	// based on the format, write the correct data to disk.
+	log.Info("inspecting restore file to determine header")
+	checksum := sha256.New()
+	sr, closer, err := snapshotReader(ri.StagePath, checksum)
+	if err != nil {
+		log.Error("error while opening restore file", "err", err)
+		return db, err
+	}
+	defer closer()
+
+	// Get our header first, guaranteed first message
+	var header pb.Snapshot_Header
+	if err := sr.ReadMsg(&header); err != nil {
+		log.Error("error while parsing restore header", "err", err)
+		return db, fmt.Errorf("error reading restore header data: %s", err)
+	}
+	log.Info("snapshot header info",
+		"created_by", header.Version.Version,
+		"format", header.Format.String(),
+	)
+
+	// We currently only support bolt
+	if header.Format != pb.Snapshot_Header_BOLT {
+		return db, fmt.Errorf("invalid snapshot format (got code: %d)", header.Format)
+	}
+
+	// Open a temporary file to write our raw bolt data
+	log.Info("reading bolt information and writing it into a new bolt db")
```

Comment:
Same question here, maybe it's worth being extra explicit in the logs for where these paths are for debugging.

```suggestion
	log.Info("reading bolt information and writing it into a new bolt db", "bolt info temp path", ri.StageTempPath)
```